File: api/cs.net/ARulesXL.cs
using System;
using System.Text;

/// <summary>
/// Summary description for arulesxl
/// </summary>
public class ARulesXL: IDisposable
{
    private LogicServer ls = null;
    private bool disposed = false;

	public ARulesXL()
	{
		//
		// TODO: Add constructor logic here
		//
	}

    // Implement IDisposable.
    // Do not make this method virtual.
    // A derived class should not be able to override this method.
    public void Dispose()
    {
        Dispose(true);
        // This object will be cleaned up by the Dispose method.
        // Therefore, you should call GC.SupressFinalize to
        // take this object off the finalization queue 
        // and prevent finalization code for this object
        // from executing a second time.
        GC.SuppressFinalize(this);
    }

    // Dispose(bool disposing) executes in two distinct scenarios.
    // If disposing equals true, the method has been called directly
    // or indirectly by a user's code. Managed and unmanaged resources
    // can be disposed.
    // If disposing equals false, the method has been called by the 
    // runtime from inside the finalizer and you should not reference 
    // other objects. Only unmanaged resources can be disposed.
    private void Dispose(bool disposing)
    {
        // Check to see if Dispose has already been called.
        if (!this.disposed)
        {
            // If disposing equals true, dispose all managed 
            // and unmanaged resources.
            if(disposing)
            {
                // Dispose managed resources.
            }
		 
            // Call the appropriate methods to clean up 
            // unmanaged resources here.
            // If disposing is false, 
            // only the following code is executed.
            if (ls != null)
            {
                ls.Close();
                ls = null;
            }
        }
        disposed = true;         
    }

    // Use C# destructor syntax for finalization code.
    // This destructor will run only if the Dispose method 
    // does not get called.
    // It gives your base class the opportunity to finalize.
    // Do not provide destructors in types derived from this class.
    ~ARulesXL()      
    {
        // Do not re-create Dispose clean-up code here.
        // Calling Dispose(false) is optimal in terms of
        // readability and maintainability.
        Dispose(false);
    }

    public void OpenRules(string rulesetfile, string arulespath)
    {
        string s;
        long term;

        try
        {
            // Get our logicserver and initialize it
            ls = new LogicServer();
            ls.Init(DoubleSlashes(arulespath + "arulesrt.cfg"));

            // Load the arulesxl engine
            ls.Load(DoubleSlashes(arulespath + "arulesrt.xpl"));

            // Load the ruleset
            if (rulesetfile != null)
            {
                s = "consult(`" + DoubleSlashes(rulesetfile) + "`)";
                term = ls.ExecStr(s);
                if (term == 0)
                    throw new Exception("Unable to consult ruleset");
            }
        }
        catch (LSException ex)
        {
            throw new Exception("OpenRules / " + FormatLSException(ex));
        }
    }

    public void CloseRules()
    {
        if (ls != null)
        {
            ls.Close();
            ls = null;
        }
    }

    public LogicServer GetLS()
    {
        return ls;
    }
    
    public string QueryRules(string ruleset, string query)
    {
        string s;
        int term, answer;
        int len;

        try
        {
            s = "arxl_query(" + ruleset + ", false, `" + query + "`, ?answer)";
            term = ls.ExecStr(s);
            if (term == 0)
                throw new Exception(FormatARulesError());

            // The answer is a string in the 4th argument
            // Could be any type, turn it into a string
            answer = ls.GetArg(term, 4);
            len = ls.StrTermLen(answer);
            return(ls.TermToStr(answer, len+1));
        }
        catch (LSException ex)
        {
            throw new Exception("Error getting arules error: " + ex.GetMsg());
        }
    }

    public string QueryMore(string ruleset, string query)
    {
        string s;
        int term, answer;
        int len;

        try
        {
            s = "arxl_query(" + ruleset + ", true, `" + query + "`, ?answer)";
            term = ls.ExecStr(s);
            if (term == 0)
                throw new Exception(FormatARulesError());

            // The answer is a string in the 4th argument
            // Could be any type, turn it into a string
            answer = ls.GetArg(term, 4);
            len = ls.StrTermLen(answer);
            return(ls.TermToStr(answer, len+1));
        }
        catch (LSException ex)
        {
            throw new Exception("Error getting arules error: " + ex.GetMsg());
        }
    }

    public void ClearTable(string ruleset, string objectname)
    {
        int term;

        term = ls.ExecStr("arxl_initialize_table(" + ruleset + ", `" + objectname + "`)");
        if (term == 0) 
            throw new Exception(FormatARulesError());
    }

    public void ClearVector(string ruleset, string objectname)
    {
        int term;

        term = ls.ExecStr("arxl_initialize_table(" + ruleset + ", `" + objectname + "`)");
        if (term == 0) 
            throw new Exception(FormatARulesError());
    }

    public void AddToTable(string ruleset, string objectname, string rowname, string colname, string value)
    {
        int term;

        term = ls.ExecStr("arxl_add_to_table(" + ruleset + ", `" + objectname + "`, `" + rowname + "`, `" + colname + "`, `" + value + "`)");
        if (term == 0) 
            throw new Exception(FormatARulesError());
    }

    public void AddToVector(string ruleset, string objectname, string rowname, string value)
    {
        int term;

        term = ls.ExecStr("arxl_add_to_vector(" + ruleset + ", `" + objectname + "`, `" + rowname + "`, `" + value + "`)");
        if (term == 0)
            throw new Exception(FormatARulesError());
    }

    public void AddObject(string ruleset, string objectname, string value)
    {
        int term;

        term = ls.ExecStr("arxl_add_data_cell(" + ruleset + ", `" + objectname + "`, `" + value + "`)");
        if (term == 0)
            throw new Exception(FormatARulesError());
    }

	public int ARulesRegisterRuntime(uint method, string proxylist, string runtimeid)
	{
		return ls.ARulesRegisterRuntime(method, proxylist, runtimeid);
	}

   // Display errors from ARulesXL
    private string FormatARulesError() 
    {
        int term;

        try
        {
            term = ls.ExecStr("query_error(?x)");
            if (term != 0)
                return(ls.GetStrArg(term, 1));
            else
                return("Unknown ARulesXL error");
        }
        catch (LSException ex)
        {
            return ("Error getting ARulesXL error: " + ex.GetMsg());
        }
    }

    // Display errors from the underlying Amzi! Logic Server
    private string FormatLSException(LSException e)
    {
        string msg = "";
        string lineno;

        try
        {
             // Consult errors
            if (e.GetExceptType() == (int)exLSTYPE.READ)
             {
                lineno = e.GetReadLineno().ToString();
                msg = e.GetMsg() + "\nin file " + e.GetReadFileName() + "\nat line " + lineno + "\n" + e.GetReadBuffer();
                return (msg);
             }

             // Other Logic Server error
             msg = e.GetMsg();
             return (msg);
        }
        catch (LSException)
        {
            return ("Error catching ARulesXL error");
        }
    }

    private string DoubleSlashes(string path)
    {
      int i;
      StringBuilder sb;

      sb = new StringBuilder(path);
      for (i = 0 ; i < sb.Length ; i++) 
      {
         if (sb[i] == '\\') 
         {
            sb.Insert(i, '\\');
            i++;
         }
      }
      return(sb.ToString());
    }

}

File: api/delphi/arulesxl.pas
unit ARulesXL;

{
  ARulesXL.PAS -- ARulesXL Delphi API Calls the Amzi! Logic Server
  Copyright (c) 2006 Amzi! inc. All Rights Reserved.
}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Arules;

type
  EARulesXLException = class(Exception);

  TARulesXL = class(TComponent)
  private
    ls: TLSARules;
    initialized: Boolean;
    function formatARulesError(): string;
    function formatLSException(): string;
  protected
  public
    constructor Create(Owner: TComponent); override;
    destructor Destroy; override;

    function GetLS(): TLSARules;

    { ARulesXL Standard Methods }
    procedure OpenRules(rulesfile: string; arulespath: string);
    procedure CloseRules();

    procedure ClearTable(ruleset: string; objectname: string);
    procedure ClearVector(ruleset: string; objectname: string);
    procedure AddToTable(ruleset: string; objectname: string; rowname: string; colname: string; value: string); overload;
    procedure AddToTable(ruleset: string; objectname: string; rownum: integer; colnum: integer; value: string); overload;
    procedure AddToTable(ruleset: string; objectname: string; rowname: string; colname: string; ivalue: integer); overload;
    procedure AddToTable(ruleset: string; objectname: string; rownum: integer; colnum: integer; ivalue: integer); overload;
    procedure AddToVector(ruleset: string; objectname: string; rowname: string; value: string); overload;
    procedure AddToVector(ruleset: string; objectname: string; rownum: integer; value: string); overload;
    procedure AddToVector(ruleset: string; objectname: string; rowname: string; ivalue: integer); overload;
    procedure AddToVector(ruleset: string; objectname: string; rownum: integer; ivalue: integer); overload;
    procedure AddObject(ruleset: string; objectname: string; value: string); overload;
    procedure AddObject(ruleset: string; objectname: string; ivalue: integer); overload;

    function QueryRules(ruleset: string; query: string): string;
    function QueryMore(ruleset: string; query: string): string;

    function ARulesRegisterRuntime(method: longint; proxylist: string; runtimeid: string): integer;

    { Delphi-Only Methods }
    function QueryRulesToStringList(ruleset: string; query: string; answer: TStringList): boolean;
    function QueryMoreToStringList(ruleset: string; query: string; answer: TStringList): boolean;
    procedure LoadTableFromNameValueList(ruleset: string; objectname: string; col1name: string; col2name: string; nvlist: TStrings);

  published
  end;

procedure Register;

implementation

constructor TARulesXL.Create(Owner: TComponent);
begin
  inherited Create(Owner);
  ls := TLSARules.Create(Owner);
  initialized := false;
end;

destructor TARulesXL.Destroy;
begin
  if initialized then ls.Close();
  inherited Destroy;
end;

function TARulesXL.GetLS(): TLSARules;
begin
  Result := ls;
end;

procedure TARulesXL.OpenRules(rulesfile: string; arulespath: string);
var
  tf: boolean;
  t: TTerm;
begin
  try
    ls.InitLS(arulespath + 'arulesrt');  { Initialize engine }
    initialized := true;
    ls.Load(arulespath + 'arulesrt'); { Load .xpl file }
    tf := ls.ExecPStr(t, 'load(''' + rulesfile + ''')');
    if (tf <> true) then
      raise EARulesXLException.Create('Unable to load rules file: '+rulesfile);
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

procedure TARulesXL.CloseRules();
begin
  try
    ls.Close();
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
  initialized := false;
end;

procedure TARulesXL.ClearTable(ruleset: string; objectname: string);
var
  t: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_initialize_table(' + ruleset +', `' + objectname + '`)');
    if (tf = false) then
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

procedure TARulesXL.ClearVector(ruleset: string; objectname: string);
begin
  ClearTable(ruleset, objectname);
end;

procedure TARulesXL.AddObject(ruleset: string; objectname: string; value: string);
var
  t: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_add_data_cell(' + ruleset + ', `' + objectname +
      '`, `' + value + '`)');
    if (tf = false) then
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

procedure TARulesXL.AddObject(ruleset: string; objectname: string; ivalue: integer);
var
  value: string;
begin
  Str(ivalue, value);
  AddObject(ruleset, objectname, value);
end;

procedure TARulesXL.AddToVector(ruleset: string; objectname: string; rowname: string; value: string);
var
  t: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_add_to_vector(' + ruleset + ', `' + objectname +
      '`, `' + rowname + '`, `' + value + '`)');
    if (tf = false) then
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

procedure TARulesXL.AddToVector(ruleset: string; objectname: string; rownum: integer; value: string);
var
  rowname: string;
begin
  Str(rownum, rowname);
  AddToVector(ruleset, objectname, rowname, value);
end;

procedure TARulesXL.AddToVector(ruleset: string; objectname: string; rowname: string; ivalue: integer);
var
  value: string;
begin
  Str(ivalue, value);
  AddToVector(ruleset, objectname, rowname, value);
end;

procedure TARulesXL.AddToVector(ruleset: string; objectname: string; rownum: integer; ivalue: integer);
var
  rowname, value: string;
begin
  Str(rownum, rowname);
  Str(ivalue, value);
  AddToVector(ruleset, objectname, rowname, value);
end;

procedure TARulesXL.AddToTable(ruleset: string; objectname: string; rowname: string; colname: string; value: string);
var
  t: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_add_to_table(' + ruleset + ', `' + objectname +
      '`, `' + rowname + '`, `' + colname + '`, `' + value + '`)');
    if (tf = false) then
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

procedure TARulesXL.AddToTable(ruleset: string; objectname: string; rowname: string; colname: string; ivalue: integer);
var
  value: string;
begin
  Str(ivalue, value);
  AddToTable(ruleset, objectname, rowname, colname, value);
end;

procedure TARulesXL.AddToTable(ruleset: string; objectname: string; rownum: integer; colnum: integer; value: string);
var
  rowname, colname: string;
begin
  Str(rownum, rowname);
  Str(colnum, colname);
  AddToTable(ruleset, objectname, rowname, colname, value);
end;

procedure TARulesXL.AddToTable(ruleset: string; objectname: string; rownum: integer; colnum: integer; ivalue: integer);
var
  rowname, colname, value: string;
begin
  Str(rownum, rowname);
  Str(colnum, colname);
  Str(ivalue, value);
  AddToTable(ruleset, objectname, rowname, colname, value);
end;

function TARulesXL.QueryRules(ruleset: string; query: string): string;
var
  t, answer: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_query(' + ruleset + ', false, `' + query + '`, ?x)');
    if (tf) then
    begin
      ls.GetArg(t, 4, dTERM, @answer);
      Result := ls.TermToPStr(answer);
    end
    else
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

function TARulesXL.QueryMore(ruleset: string; query: string): string;
var
  t, answer: TTerm;
  tf: boolean;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_query(' + ruleset + ', true, `' + query + '`, ?x)');
    if (tf) then
    begin
      ls.GetArg(t, 4, dTERM, @answer);
      Result := ls.TermToPStr(answer);
    end
    else
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

{ Delphi-Only Functions }

procedure TARulesXL.LoadTableFromNameValueList(ruleset: string; objectname: string; col1name: string; col2name: string; nvlist: TStrings);
var
  t: TTerm;
  tf: boolean;
  i: integer;
  name, value, idx: string;
begin
  try
    tf := ls.ExecPStr(t, 'arxl_initialize_table(' + ruleset + ', `' + objectname + '`)');
    if (tf <> true) then
      raise EARulesXLException.Create(FormatARulesError());

    for i := 0 to nvlist.Count-1 do
    begin
      Str(i, idx);
      name := nvlist.Names[i];
      value := nvlist.ValueFromIndex[i];
      tf := ls.ExecPStr(t, 'arxl_add_to_table(' + idx +  ', `' + col1name + '`, `' + name + '`)');
      if (tf = false) then
        raise EARulesXLException.Create(FormatARulesError());
      tf := ls.ExecPStr(t, 'arxl_add_to_table(' + idx +  ', `' + col2name + '`, `' + value + '`)');
      if (tf = false) then
        raise EARulesXLException.Create(FormatARulesError());
    end;
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;
end;

function TARulesXL.QueryRulesToStringList(ruleset: string; query: string; answer: TStringList): boolean;
var
  t, list, item: TTerm;
  argtype: TPType;
  s: string;
  tf: boolean;
  rc: integer;
  msg: array [0..1000] of char;
begin
  try
    { Query the Logic Server }
    tf := ls.ExecPStr(t, 'arxl_query(' + ruleset + ', false, `' + query + '`, ?x)');

    if (tf) then
      begin

      { Get the 4th argument of the result back }
      argtype := ls.GetArgType(t, 4);
      ls.GetArg(t, 4, dTERM, @list);
      if (argtype <> pLIST) then
      begin
        ls.TermToStr(list, msg, 1000);
        raise EARulesXLException.Create(msg);
      end;

      { Walk the list of lists }
      repeat
        rc := ls.PopList(list, dTERM, @item);
        if (rc = 0) then
        begin
          s := ls.TermToPStr(item);
          answer.Add(s);
        end
      until (rc <> 0);
      end
    else
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;

  Result := tf;
end;

function TARulesXL.QueryMoreToStringList(ruleset: string; query: string; answer: TStringList): boolean;
var
  t, list, item: TTerm;
  argtype: TPType;
  s: string;
  tf: boolean;
  rc: integer;
  msg: array [0..1000] of char;
begin
  try
    { Query the Logic Server }
//  before := Time;
    tf := ls.ExecPStr(t, 'arxl_query(' + ruleset + ', true, `' + query + '`, ?x)');
{  after := Time;
  elapsed := MilliSecondsBetween(before, after);
  Str(elapsed, s);
  ShowMessage('Time= ' + s);
}
    if (tf) then
      begin

      { Get the 4th argument of the result back }
      argtype := ls.GetArgType(t, 4);
      ls.GetArg(t, 4, dTERM, @list);
      if (argtype <> pLIST) then
      begin
        ls.TermToStr(list, msg, 1000);
        raise EARulesXLException.Create(msg);
      end;

      { Walk the list of lists }
      repeat
        rc := ls.PopList(list, dTERM, @item);
        if (rc = 0) then
        begin
          s := ls.TermToPStr(item);
          answer.Add(s);
        end
      until (rc <> 0);
      end
    else
      raise EARulesXLException.Create(FormatARulesError());
  except on ELogicServer do raise EARulesXLException.Create(formatLSException());
  end;

  Result := tf;
end;

function TARulesXL.ARulesRegisterRuntime(method: longint; proxylist: string; runtimeid: string): integer;
begin
  Result := ls.ARulesRegisterRuntime(method, proxylist, runtimeid);
end;

function TARulesXL.FormatARulesError(): string;
var
  t: TTerm;
  tf: boolean;
  msg: array [0..1000] of char;
begin
  try
    tf := ls.ExecPStr(t, 'query_error(?x)');
    If tf = true Then
      Result := ls.GetPStrArg(t, 1)
    Else
      Result := 'Unknown ARulesXL error'
  except on ELogicServer do
    begin
    ls.GetExceptMsg(msg, 1000);
    Result := 'Error getting ARulesXL error: ' + msg;
    end;
  end;
end;

function TARulesXL.FormatLSException(): string;
var
  msg: array [0..1000] of char;
begin
  try
    ls.GetExceptMsg(msg, 1000);
    Result := 'Logic Server Error: ' +  msg;
  except on ELogicServer do Result := 'Error getting Logic Server error';
  end;
end;


procedure Register;
begin
  RegisterComponents('Additional', [TARulesXL]);
end;

end.
